src/abe_policy/access_policy.rs

```diff
@@ -1,430 +1,239 @@
+/// An access policy is a boolean expression of attributes.
+///
+/// TODO: is this a subset-cover limitation? It seems possible to subtract
+/// coordinates from the set of positively generated ones.
 ///
 /// Only `positive` literals are allowed (no negation).
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub enum AccessPolicy {
     Attr(Attribute),
     And(Box<AccessPolicy>, Box<AccessPolicy>),
     Or(Box<AccessPolicy>, Box<AccessPolicy>),
-    All, // indicates we want the disjunction of all attributes
-}
-
-impl PartialEq for AccessPolicy {
-    fn eq(&self, other: &Self) -> bool {
-        let mut attributes_mapping = HashMap::<Attribute, u32>::new();
-        let left_to_u32 = self.to_u32(&mut attributes_mapping);
-        let right_to_u32 = other.to_u32(&mut attributes_mapping);
-        if left_to_u32 != right_to_u32 {
-            false
-        } else {
-            self.attributes() == other.attributes()
-        }
-    }
+    Any,
 }
 
 impl AccessPolicy {
-    /// Creates an Access Policy based on a single Policy Attribute.
-    ///
-    /// Shortcut for
-    /// ```ignore
-    /// AccessPolicy::Attr(Attribute::new(dimension, attribute))
-    /// ```
-    ///
-    /// Access Policies can easily be created using it
-    /// ```ignore
-    /// let access_policy =
-    ///     AccessPolicy::new("Security Level", "level 4")
-    ///         & (AccessPolicy::new("Department", "MKG") | AccessPolicy::new("Department", "FIN"));
-    /// ```
-    #[must_use]
-    pub fn new(dimension: &str, attribute: &str) -> Self {
-        Self::Attr(Attribute::new(dimension, attribute))
-    }
-
-    /// Converts policy to integer value (for comparison).
-    /// Each attribute is mapped to an integer value and the algebraic
-    /// expression is applied with those values.
-    /// We must keep a mapping of each attribute to the corresponding integer
-    /// value in order to avoid having 2 different attributes with same integer
-    /// value
-    fn to_u32(&self, attribute_mapping: &mut HashMap<Attribute, u32>) -> u32 {
-        match self {
-            Self::Attr(attr) => {
-                if let Some(integer_value) = attribute_mapping.get(attr) {
-                    *integer_value
-                } else {
-                    let max = (attribute_mapping.len() + 1) as u32;
-                    attribute_mapping.insert(attr.clone(), max);
-                    max
-                }
-            }
-            Self::And(l, r) => l.to_u32(attribute_mapping) * r.to_u32(attribute_mapping),
-            Self::Or(l, r) => l.to_u32(attribute_mapping) + r.to_u32(attribute_mapping),
-            Self::All => 0,
-        }
-    }
-
-    /// This function is finding the right closing parenthesis in the boolean
-    /// expression given as a string
+    /// Finds the corresponding closing parenthesis in the boolean expression
+    /// given as a string.
     fn find_next_parenthesis(boolean_expression: &str) -> Result<usize, Error> {
```

Comment:
The name of the function does not clearing indicate that we are looking for a closing parenthesis.

Reply:
done in 314c441